src/System.Private.CoreLib/shared/System/MemoryExtensions.Fast.cs

```diff
@@ -400,6 +400,53 @@ public static Span<T> AsSpan<T>(this T[] array, int start)
             return new Span<T>(ref Unsafe.Add(ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData()), start), array.Length - start);
         }
 
+        /// <summary>
+        /// Creates a new span over the portion of the target array.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Span<T> AsSpan<T>(this T[] array, Index startIndex)
+        {
+            if (array == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            }
+
+            if (default(T) == null && array.GetType() != typeof(T[]))
+                ThrowHelper.ThrowArrayTypeMismatchException();
+
+            int actualIndex = startIndex.IsFromEnd ? array.Length - startIndex.Value : startIndex.Value;
+
+            if ((uint)actualIndex > (uint)array.Length)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            return new Span<T>(ref Unsafe.Add(ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData()), actualIndex), array.Length - actualIndex);
+        }
+
+        /// <summary>
+        /// Creates a new span over the portion of the target array.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Span<T> AsSpan<T>(this T[] array, Range range)
+        {
+            if (array == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            }
+
+            if (default(T) == null && array.GetType() != typeof(T[]))
+                ThrowHelper.ThrowArrayTypeMismatchException();
+
+            int start = range.Start.IsFromEnd ? array.Length - range.Start.Value : range.Start.Value;
+            int end = range.End.IsFromEnd ? array.Length - range.End.Value : range.End.Value;
```

Comment:
This logic is duplicated in several places. Can we centralize it? And if it was deemed useful to have GetOffset for Index, should this equivalent be exposed for Range?

Reply:
We already exposing GetOffsetLength from Range but I avoided calling this one to bypass the validation. But I can see it is a good idea we can have a version bypass the extra check and use it. I am not sure if this can slow the jit or not though. @AndyAyersMS what you think?

Reply:
Thanks, I hadn't seen that yet.  If it's not good enough for our own use, we still believe it'll be good enough for others' use?

Reply:
It depends on the callers scenarios. if it is perf critical they can still do the calculation manually as we do. if not, then this method will be very convenient.